lib/src/sdk_env.dart

```diff
@@ -276,15 +276,33 @@ class ToolEnvironment {
     return false;
   }
 
-  Future<ProcessResult> runUpgrade(String packageDir, bool usesFlutter,
-      {int retryCount = 3}) async {
-    final backup = await _stripPubspecYaml(packageDir);
+  FutureOr<T> _withTempDir<T>(FutureOr<T> Function(Directory) fn) async {
+    var tempDir = Directory.systemTemp.createTempSync('pana_');
     try {
+      return await fn(tempDir);
+    } finally {
+      if (tempDir.existsSync()) {
+        tempDir.deleteSync(recursive: true);
+      }
+    }
+  }
+
+  /// Runs `pub upgrade` on a stripped copy of [pubspec].
+  Future<ProcessResult> runUpgrade(
```

Comment:
This is being called from `pub-dev`'s `dartdoc_runner.dart` too, before running `dartdoc`. Is copying the `pubspec.lock` enough to support that?

Reply:
Not sure.

Are we using the resolution afterwards then this will probably not be enough.
If we just want the fact that it resolved then this should be enough. (Actually we probably do not want to copy a pubspec.lock file when running on a hosted package, but rely on an open-ended resolution).

Reply:
I'd rather change the local package analysis path to copy the entire package to a temp directory instead of using thing - otherwise nice - trick...

Reply:
I don't think this is very much a trick. Not more than the old `_stripPubspecYaml` method.
If we copy the entire folder we still have to think about the interaction with`dartdoc_runner.dart` from `pub-dev`.
So the question remains:
Does `dartdoc_runner.dart` rely on the side-effects of running the pub-upgrade?

Reply:
> Does dartdoc_runner.dart rely on the side-effects of running the pub-upgrade?

The reason we run this before `dartdoc` is to have a full package setup (so that `dartdoc` doesn't need to run it), in order to separate the timeouts of fetching the dependencies from the already-long dartdoc content generation. I don't recall the details, but for a while this also fixed an issue where `dartdoc` failed to get some Flutter dependencies properly.

> If we copy the entire folder we still have to think about the interaction with dartdoc_runner.dart from pub-dev.

My thinking was:
- let's leave this method as-is (after all, it is called run upgrade)
- for the local analysis path let's prepare the analysis with copying the package to a temp dir, and run every destructive modification on the temp copy.

Reply:
I will leave the method as-is, so dartdocRunner can continue working.
But pana itself will not copy the entire directory.
Though I think moving forward we should let dartdoc-runner do its own pub upgrade.

Does it work for you?